server/src/main/java/org/elasticsearch/index/search/stats/ShardFieldUsageTracker.java

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.index.search.stats;
+
+import org.elasticsearch.core.Releasable;
+import org.elasticsearch.index.search.stats.FieldUsageStats.PerFieldUsageStats;
+import org.elasticsearch.search.internal.FieldUsageTrackingDirectoryReader.FieldUsageNotifier;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class ShardFieldUsageTracker {
+
+    private final Map<String, InternalFieldStats> perFieldStats = new ConcurrentHashMap<>();
+
+    public FieldUsageStatsTrackingSession createSession() {
+        return new FieldUsageStatsTrackingSession();
+    }
+
+    public FieldUsageStats stats() {
+        final Map<String, PerFieldUsageStats> stats = new HashMap<>(perFieldStats.size());
+        for (Map.Entry<String, InternalFieldStats> entry : perFieldStats.entrySet()) {
+            PerFieldUsageStats pf = new PerFieldUsageStats();
+            InternalFieldStats ifs = entry.getValue();
+            pf.terms = ifs.terms.get();
+            pf.freqs = ifs.freqs.get();
+            pf.positions = ifs.positions.get();
+            pf.offsets = ifs.offsets.get();
+            pf.docValues = ifs.docValues.get();
+            pf.storedFields = ifs.storedFields.get();
+            pf.norms = ifs.norms.get();
+            pf.payloads = ifs.payloads.get();
+            pf.impacts = ifs.impacts.get();
+            pf.termVectors = ifs.termVectors.get();
+            pf.points = ifs.points.get();
+            stats.put(entry.getKey(), pf);
+        }
+        return new FieldUsageStats(Collections.unmodifiableMap(stats));
+    }
+
+    static class InternalFieldStats {
+        final AtomicLong terms = new AtomicLong();
+        final AtomicLong freqs = new AtomicLong();
+        final AtomicLong positions = new AtomicLong();
+        final AtomicLong offsets = new AtomicLong();
+        final AtomicLong docValues = new AtomicLong();
+        final AtomicLong storedFields = new AtomicLong();
+        final AtomicLong norms = new AtomicLong();
+        final AtomicLong payloads = new AtomicLong();
+        final AtomicLong impacts = new AtomicLong();
+        final AtomicLong termVectors = new AtomicLong();
+        final AtomicLong points = new AtomicLong();
```

Comment:
Is `LongAdder` a better fit for this use-case?

Reply:
I was worried about memory consumption of these per-shard per-field per-data-structure level objects, where AtomicLong fares better than LongAdder. Looking a bit deeper into LongAdder, it uses not that much memory when there's no contention, so assuming that most fields won't have contention, moving to LongAdder should be fine. Thanks!